File: runtime/go/lazuli/rate_limit.go
package lazuli

import "os"

// RateLimit is the env-qualified throttle policy emitted by codegen
// for each Command / Api / Agent / Report. Generated `*.gen.go`
// literals (`dist/go/<feature>/<feature>.gen.go`) populate `Default`
// from the unqualified `rate_limit "..."` source line and `ByEnv`
// from each `rate_limit "..." in <env_list>` override. The runtime
// resolves the active limit per request via `Resolve()`; the
// resolution is a single linear scan against `LAZULI_ENV`, no engine.
//
// Backward-compat: legacy single-line `rate_limit "X per Y per Z"`
// lowers to `RateLimit{Default: "X per Y per Z"}` (zero-length
// `ByEnv`); `Resolve()` returns `Default` regardless of env. Existing
// 0.15.0 fixtures emit this shape unchanged after the LZIR_SCHEMA
// bump (proposal §8).
//
// See `docs/proposals/ir-rate-limit-env-aware.md` §4.1 + §7.
type RateLimit struct {
	Default string
	ByEnv   []RateLimitByEnv
}

// RateLimitByEnv is one env-qualified override entry. `Envs` carries
// lowercase env names matching the closed catalog (`production`,
// `staging`, `test`, `dev`, `local`); `Limit` is the spec literal
// (empty string == unlimited; the proposal-defined `"unlimited"`
// keyword lowers to empty at parse time per §4.4).
type RateLimitByEnv struct {
	Envs  []string
	Limit string
}

// Resolve returns the rate-limit string active for the current
// `LAZULI_ENV`. Scans `ByEnv` linearly (catalog has at most 5
// entries; per-command lists are short) and falls back to `Default`
// when no env matches.
//
// Reads `os.Getenv("LAZULI_ENV")` per request — cheap enough not to
// cache, and lets test fixtures swap the env mid-run without
// re-instantiating handler chains. Empty `LAZULI_ENV` is treated as
// `"local"` (matching the existing CORS convention at
// `runtime/go/lazuli/http_cors.go:34`).
//
// Per RULE-VOCAB-03: zero branches on user input. The scan is a
// static equality match against the env value; no interpreter, no
// DSL evaluation.
func (r RateLimit) Resolve() string {
	env := os.Getenv("LAZULI_ENV")
	if env == "" {
		env = "local"
	}
	for _, entry := range r.ByEnv {
		for _, name := range entry.Envs {
			if name == env {
				return entry.Limit
			}
		}
	}
	return r.Default
}

// IsUnlimited reports whether the resolved limit disables throttling
// entirely (empty string == no limit). Backward-compat: the legacy
// shape `RateLimit{Default: ""}` reads as unlimited (it already did
// — the middleware skipped enforcement on empty strings before
// `ir-rate-limit-env-aware`).
func (r RateLimit) IsUnlimited() bool {
	return r.Resolve() == ""
}

// RateLimitFromDefault returns a `RateLimit` with only `Default`
// populated. Provided as a backward-compat shim for legacy callers
// that hand-constructed a string-typed `RateLimit` before the
// env-aware refactor.
//
// Generated code does not call this — codegen emits the struct
// literal directly per `format_rate_limit_struct` in
// `crates/lazuli_codegen_go/src/emitter/command.rs`.
func RateLimitFromDefault(s string) RateLimit {
	return RateLimit{Default: s}
}

File: crates/lazuli_ir/src/nodes/rate_limit.rs
//! Env-aware rate-limit specification IR.
//!
//! `ir-rate-limit-env-aware §4` — the lowered shape behind a
//! `rate_limit "<text>"` (default-only) authoring line, or the
//! multi-line `rate_limit "<text>" in <env, env>` form that overrides
//! per environment. The runtime helper `ResolveLimit()` (cell 2)
//! reads `LAZULI_ENV`, scans `by_env` in source order, and returns
//! the first matching `limit` — falling through to `default`.
//!
//! The `"unlimited"` keyword (proposal §4.4) lowers to the empty
//! string (either as the default or inside a `RateLimitByEnv.limit`)
//! so downstream consumers can treat absence of throttle uniformly.

use serde::{Deserialize, Serialize};

use crate::SpanRef;

/// ir-rate-limit-env-aware §4.1 — env-qualified rate limit container.
///
/// Backward-compat: the single-line `rate_limit "X"` source shape lowers
/// to `RateLimitSpec { default: "X", by_env: vec![] }`. The runtime helper
/// `ResolveLimit()` (cell 2) reads `LAZULI_ENV`, scans `by_env` in source
/// order, and returns the matching `limit` or falls through to `default`.
///
/// The `"unlimited"` keyword (proposal §4.4) lowers to an empty string —
/// either as the `default` (no throttle by default) or inside a
/// `RateLimitByEnv.limit` (no throttle for the listed envs).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RateLimitSpec {
    /// Default rate limit applied when no env-qualified entry matches.
    /// Empty string == no rate limit (the `"unlimited"` sentinel lowers
    /// here). The single-line `rate_limit "X"` source shape populates
    /// only this field; `by_env` stays empty.
    pub default: String,
    /// Env-qualified overrides scanned linearly at request time. Each
    /// entry covers one-or-more `EnvName`s sharing a limit string.
    /// Source-order is preserved; the first matching entry wins.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub by_env: Vec<RateLimitByEnv>,
    /// Span of the FIRST `rate_limit` line for this spec — points at the
    /// default-declaring line. Per-`by_env` entries carry their own span.
    /// Optional so synth-emitted specs without a source location work.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub span_ref: Option<SpanRef>,
}

impl RateLimitSpec {
    /// Backward-compat constructor — lifts the legacy single-string
    /// shape (or the lowered output of a one-line `rate_limit "X"`) into
    /// the new `RateLimitSpec` container. Call-sites that previously
    /// wrote `rate_limit: Some("X".to_owned())` swap to
    /// `rate_limit: Some(RateLimitSpec::from_default("X".to_owned()))`.
    pub fn from_default(s: String) -> Self {
        Self {
            default: s,
            by_env: Vec::new(),
            span_ref: None,
        }
    }
}

/// ir-rate-limit-env-aware §4.1 — single env-qualified override row.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RateLimitByEnv {
    /// Env names this entry matches. Catalog is closed (`EnvName`);
    /// identifiers outside the catalog land in `unknown_envs` and trigger
    /// the doctor diagnostic `rate_limit_unknown_env` (Cell 3).
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub envs: Vec<EnvName>,
    /// Forward-compatible bucket for env identifiers outside the closed
    /// catalog. Parses OK at AST level; Cell 3 doctor emits
    /// `rate_limit_unknown_env`. Empty for well-formed source.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub unknown_envs: Vec<String>,
    /// Limit string (e.g. `"100 per 10 minutes per ip"`) or the empty
    /// string when the source authored `"unlimited"` (proposal §4.4).
    pub limit: String,
    /// Span of this `rate_limit ... in <envs>` line in the source.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub span_ref: Option<SpanRef>,
}

/// ir-rate-limit-env-aware §4.3 — closed catalog of recognized
/// `LAZULI_ENV` values. Adding a variant is an IR change requiring a
/// proposal. The JSON form is snake_case (matches the runtime
/// `LAZULI_ENV` strings the existing CORS/session code reads).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EnvName {
    Production,
    Staging,
    Test,
    Dev,
    Local,
}

impl EnvName {
    /// Parse a lowercase identifier into the closed catalog. Returns
    /// `None` for identifiers outside the catalog; callers (the parser
    /// today, Cell 3 doctor tomorrow) decide whether to surface a
    /// warning.
    pub fn from_ident(ident: &str) -> Option<Self> {
        match ident {
            "production" => Some(EnvName::Production),
            "staging" => Some(EnvName::Staging),
            "test" => Some(EnvName::Test),
            "dev" => Some(EnvName::Dev),
            "local" => Some(EnvName::Local),
            _ => None,
        }
    }

    /// Canonical lowercase identifier (matches `LAZULI_ENV` strings).
    pub fn as_str(&self) -> &'static str {
        match self {
            EnvName::Production => "production",
            EnvName::Staging => "staging",
            EnvName::Test => "test",
            EnvName::Dev => "dev",
            EnvName::Local => "local",
        }
    }
}
